tfworker/util/system.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2020 Richard Maynard (richard.maynard@gmail.com)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import shlex
+import subprocess
+
+
+def pipe_exec(args, stdin=None, cwd=None, env=None):
+    """
+    A function to accept a list of commands and pipe them together.
+
+    Takes optional stdin to give to the first item in the pipe chain.
+    """
+    count = 0
+    commands = []
+    if env is None:
+        env = os.environ.copy()
+
+    if not isinstance(args, list):
+        args = [args]
```

Comment:
Is this to handle cases where args is a single string e.g. `"ls"` instead of `["ls"]`?

Reply:
exactly, the function is built to take it either way, when it passes it to the system though it must be a list.

Reply:
> 
> 
> exactly, the function is built to take it either way, when it passes it to the system though it must be a list.

^^ not exactly accurate, sorry each item in the list is a command + arguments, if a list with more than one element is provided it is treated a pipeline, `["ls", "wc -l"] == `ls | wc -l`